### PR TITLE
Proposal to implement Buildbot credentials management

### PR DESCRIPTION
**Pull request by protatremy**
Preliminary documentation explain planned work for supporting buildbot credential management

### CODE REVIEW

File: master/docs/manual/credentialmanagement.rst [hunk 1]
```diff
@@ -0,0 +1,100 @@
+Credential Management
+=====================
+
+Definition
+----------
+
+Buildbot needs to send and/or execute commands requiring credentials.
+
+The goal is to store credentials used by Buildbot commands.
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: Buildbot provides a common framework for steps to use credentials.

File: master/docs/manual/credentialmanagement.rst [hunk 2]
```diff
@@ -0,0 +1,100 @@
+Credential Management
+=====================
+
+Definition
+----------
+
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: one sentence per line does not mean one paragraph by sentence :)

File: master/docs/manual/credentialmanagement.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
+Credential Management
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: The file should be name cfg-credentials.rst and be acessible from the config.rst topic index

File: master/docs/manual/credentialmanagement.rst [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+=====================
+
+Definition
+----------
+
+Buildbot needs to send and/or execute commands requiring credentials.
+
+The goal is to store credentials used by Buildbot commands.
+
+Credential could not be stored in ``buildbot.config`` or ``master.config`` because the configuration is available in a source control management.
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: The framewework provides a way to configure the credentials outside of the master.cfg.
As the master.cfg is usually stored in a SCM like git, those credentials would be available to anybody.
For this reason, the credentials are stored in the buildbot database, optionally encrypted.

File: master/docs/manual/credentialmanagement.rst [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+Credential are stored in the Buildbot database.
+
+Credentials once stored are obfuscated.
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: Credentials are never visible to the normal user via logs, and thus are transmitted to the workers using the ```:class:`Obfuscated` ```

File: master/docs/manual/credentialmanagement.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+Credential case: SSH Key
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: Several use cases are addressed by this framework:

File: master/docs/manual/credentialmanagement.rst [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+A key used to send a ssh command is usually stored in a ``ssh text file`` (like id_rsa) in the :envvar:`HOME` directory.
+
+The ssh key registering is done with 3 entities: a name (ssh key name), a value (ssh_key value) and a path (where the file is stored).
+
+Credential case: Password
+-------------------------
+
+A password could be use in a command by Buildbot.
+The password is stored with 2 entities: a ``name``, and a ``value``. No path is inserted.
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: No path is required for this usecase.

File: master/docs/manual/credentialmanagement.rst [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    .. code-block:: bash
+
+        buildbot add-creds http://localhost:8020 --file ~/.ssh/id_rsa --destination '$HOME/.ssh/id_rsa' --name ssh_user
+        buildbot add-creds http://localhost:8020  --name userpassword --value mypassword
+
+A data API allows users to modify credential values
+
+    .. code-block:: bash
+
+        buildbot modify-creds http://localhost:8020 --file ~/.ssh/id_rsa --destination '$HOME/.ssh/id_rsa' --name ssh_user
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: add and modify creds are the same. We should rather have a set-creds command which does both use cases

File: master/docs/manual/credentialmanagement.rst [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+Registering credentials through a Buildbot UI
+---------------------------------------------
+
+A Buildbot UI plugin helps to enter, modify and delete the credentials.
+
+Anywhere those credentials are needed, there is a drop down list of the appropriate available credential names, and you just select the appropriate one.
+
+When it's time to change the password, you just change it once or delete it.
+
+Credentials table database description
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: This part should be rather documented in the db.rst

File: master/docs/manual/credentialmanagement.rst [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  +--------------------+-------+----------------+
+  |   pass_user1       | XXXXX |                |
+  +--------------------+-------+----------------+
+
+How to use credentials stored in the database
+---------------------------------------------
+
+Credentials are stored in the database and have to be use during a build.
+
+A DB API function helps to populate the credentials in a master build step.
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: implementation details not required in the user manual

File: master/docs/manual/credentialmanagement.rst [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+Credentials are interpolated in the build like properties are but in a first step, and then deleted at the end of the build.
```
Comment (tardyp) [inline master/docs/manual/credentialmanagement.rst]: You can use the normal Interpolate function to use password credentials in a ShellCommand. In this case, you would use the `creds` keyword of Interpolate. (This needs to be documented also in the Interpolate doc)

The second file credential usecase is implemented using two Steps: PopulateCreds and RemoveCreds, which can be used as follow::

File: master/docs/manual/credentialmanagement.rst (hunk 1 above)
Comment (tomprince) [inline master/docs/manual/credentialmanagement.rst]: @tardyp I don't think this document is not intended to be included in the docs as-is, but is rather aimed at the developers to get feedback on the idea. Thus responding to specific details of the wording, formatting and organization in the documentation probably isn't relevant at this point.

**Comment by tomprince** [comment]
I think having a story for credential management is a good idea. However, I don't think buildbot should get into the business of managing secrets itself. Rather I think it would make more sense to integrate with existing systems for that (e.g. [Vault](https://www.hashicorp.com/blog/vault.html), [Barbican](https://wiki.openstack.org/wiki/Barbican)) rather than trying to get secure credential storage right.

**Comment by tardyp** [comment]
@tomprince I think this is a good idea, but I would prefer to make them an optional backend.

An unsecure db backend would help user to start simple, and then scale by using/implementing more complex backend, while still keeping their  Interpolated creds config

**Comment by tardyp** [comment]
full disclosure, remy works just at my side cubicle.

**Comment by tomprince** [comment]
> An unsecure db backend would help user to start simple,

I think having an insecure default would be a disservice to users.

**Comment by tardyp** [comment]
> I think having an insecure default would be a disservice to users.

For now, I have all my buildbot_travis users which are storing their creds in the environment variables. So Having a writeonly credential store is much much better.

**Comment by tardyp** [comment]
After more thought, and looking at vault, I agree that indeed a naive implementation using the buildbot db is not something desirable. I think we need a proper interface layer in order to abstract vault/barbican api.

It looks so easy to install vault the the user who care about not to expose creds to the conf will always care about storing them properly, so there is not so much usecase for the intermediate solution.


Then there is still the bootstrap problem on how to store/configure the api token for vault/barbican.
This can be left as a single configuration with secure storage problem left to the user.

**Comment by protatremy** [comment]
Ok, I take some times to understand the documentation and try some tricks with Vault and Barbican, and I will submit a new documentation ASAP, including this solution.

File: master/docs/manual/credentialmanagement.rst [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+--------------------------------------------
+
+A data API is available to create, delete and modify new credentials.
+
+The data API modifies the 3 entities:
+``NAME`` and ``VALUE`` (mandatory), ``PATH`` (not mandatory, as it is not used in the password case).
+
+A data API function helps to register any new credential in the database.
+
+    .. code-block:: bash
```
Comment (sa2ajj) [inline master/docs/manual/credentialmanagement.rst]: no need for extra indentation here (including the code block lines)

File: master/docs/manual/credentialmanagement.rst [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+A password could be use in a command by Buildbot.
+The password is stored with 2 entities: a ``name``, and a ``value``. No path is inserted.
+
+How to add, delete or modify new credentials
+--------------------------------------------
+
+A data API is available to create, delete and modify new credentials.
+
+The data API modifies the 3 entities:
+``NAME`` and ``VALUE`` (mandatory), ``PATH`` (not mandatory, as it is not used in the password case).
```
Comment (sa2ajj) [inline master/docs/manual/credentialmanagement.rst]: these might benefit of being a simple list

File: master/docs/manual/credentialmanagement.rst [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+A data API function helps to register any new credential in the database.
+
+    .. code-block:: bash
+
+        buildbot add-creds http://localhost:8020 --file ~/.ssh/id_rsa --destination '$HOME/.ssh/id_rsa' --name ssh_user
+        buildbot add-creds http://localhost:8020  --name userpassword --value mypassword
+
+A data API allows users to modify credential values
+
+    .. code-block:: bash
```
Comment (sa2ajj) [inline master/docs/manual/credentialmanagement.rst]: ditto

File: master/docs/manual/credentialmanagement.rst [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+        buildbot modify-creds http://localhost:8020 --file ~/.ssh/id_rsa --destination '$HOME/.ssh/id_rsa' --name ssh_user
+        buildbot modify-creds http://localhost:8020  --name userpassword --value mypassword
+
+A data API allows users to delete credentials
+
+    .. code-block:: bash
```
Comment (sa2ajj) [inline master/docs/manual/credentialmanagement.rst]: ditto

**Comment by sa2ajj** [comment]
I think a bit of code that implements what documentation talks about would be really nice: I agree that docs upfront is a good idea, however code + docs help to better understand both sides.

**Comment by sa2ajj** [comment]
This is one of the topics I've been pondering on: there are, at least, two different scenarios -- create an instance of something that requires a credential (e.g. build worker); run a command that requires a secret of sort (e.g. `scp` something using a particular SSH key).

What I'd like to see is something that would cover both use cases.  This particular proposal does not seem to offer that.

File: master/docs/manual/cfg-credentials.rst [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+Vault to store credentials
+--------------------------
+
+Vault secures, stores, and tightly controls access to credentials. Vault presents an unified API to access multiple backends.
+Builbot instance need a key and a token to be authentificated by a Vault server.
+Vault store tuples (key, value).
+An interface between Vault and Buildbot helps to acces to the credentials versus API commands.
+
+How to add, delete or modify new credentials
```
Comment (tardyp) [inline master/docs/manual/cfg-credentials.rst]: Is it really necessary?
The vault command line cannot be used for that purpose?

File: master/docs/manual/cfg-credentials.rst [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+A Generic API function helps to populate the credentials in a master build step, getting the credentials in Vault backend.
+Credentials are interpolated in the build like properties are but in a first step, and then deleted at the end of the build.
+The class PopulateCreds take two kwargs arguments, ssh_keys and passwords.
+When ssh_keys argument is not empty, a file ``id_rsa_credential_name`` is created for each ssh key needed in the :envvar:`HOME` directory.
+Credentials values are obfuscated in the steps logs.
+
+.. code-block:: python
+
+        # e.g in abuild:
+        f1.addStep(PopulateCreds(ssh_keys=['ssh_user'], passwords=['userpassword'])
```
Comment (tardyp) [inline master/docs/manual/cfg-credentials.rst]: where would you populate passwords?

I think it is not a good idea to have a specific hardcoded path for ssh keys. We should rather have something like:

 f1.addStep(PopulateCreds([CredsWithPath(name='ssh_user', path="~/.ssh/id_rsa")]))

CredsWithPath being just a named tuple

File: master/docs/manual/cfg-credentials.rst [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+        f1.addStep(ShellCommand(Interpolate("wget -u user -p %{creds:userpassword}s %{prop:urltofetch}s")))
+        f1.addStep(RemoveCreds(['ssh_user'])
```
Comment (tardyp) [inline master/docs/manual/cfg-credentials.rst]: How will RemoveCreds remember where to delete the 'ssh_user' key?

File: master/docs/manual/cfg-credentials.rst [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+.. code-block:: bash
+
+    buildbot delete-creds http://localhost:8020  --name ssh_user
+    buildbot delete-creds http://localhost:8020  --name userpassword
+
+Registering credentials through a Buildbot UI
+---------------------------------------------
+
+A Buildbot UI plugin helps to enter, modify and delete the credentials.
```
Comment (tardyp) [inline master/docs/manual/cfg-credentials.rst]: I think if we decide to use vault, we should stick with its own UI to let the user populate the creds.

Comment (tomprince) [inline master/docs/manual/cfg-credentials.rst]: > I think if we decide to use vault, we should stick with its own UI to let the user populate the creds.

:+1:

**Comment by tardyp** [comment]
I agree with @sa2ajj that this would be great to have the Vault interface to be used to store the various master side creds that are needed (e.g creds to EC2 for ec2 latent build slaves, creds to gerrit for the ssh review, creds to github, etc)

File: master/docs/manual/cfg-credentials.rst [hunk 20]
```diff
@@ -0,0 +1,79 @@
+Credential Management
+=====================
+
+Definition
+----------
+
+Buildbot needs to send and/or execute commands requiring credentials.
+The goal is to store credentials used by Buildbot commands.
+Buildbot provides a common framework for steps to use credentials.
+The framewework provides a way to configure the credentials outside of the master.cfg.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-credentials.rst]:
```
:file:`master.cfg`
```

File: master/docs/manual/cfg-credentials.rst [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+As the master.cfg is usually stored in a SCM like git, those credentials would be available to anybody.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-credentials.rst]: ditto

File: master/docs/manual/cfg-credentials.rst [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+For this reason, the credentials are stored in a Vault instance. Vault have to be installed to use this feauture.
+For more informations about Vault, please visit: _`Vault`: https://www.vaultproject.io/
+Credentials are never visible to the normal user via logs, and thus are transmitted to the workers using the :class:`Obfuscated`
+
+Several use cases are addressed by this framework:
+
+Credential case: SSH Key
+------------------------
+
+A key used to send a ssh command is usually stored in a ``ssh text file`` (like id_rsa) in the :envvar:`HOME` directory.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-credentials.rst]:
```
:file:`id_rsa`
```

Comment (sa2ajj) [inline master/docs/manual/cfg-credentials.rst]: SSH

**Comment by sa2ajj** [comment]
Any updates to the docs/code?

**Comment by protatremy** [comment]
Hi sorry for the delay, I had a baby last month and I was in holidays to help the mom. I started to work again on the credentials projects, I will upload the new doc with an architecture based on Vault and Consul. I tested the installation with a docker file on my side. I am working on the interface code between Vault and Buildbot. I hope you will be interested by this solution.

**Comment by tardyp** [comment]
superseed by https://github.com/buildbot/buildbot/pull/2660

**Pull request closed without merging**